(#117) Feature/sy alert

ebPark9511 opened the pull request:
## 💡 이성프로필 관련 커스텀 팝업

#117 이성프로필에서 사용되는 알럿을 작업했습니다.

✅ 작업 내용 

- #131 
- #132  
- #133 

---
* #130 -> 해당 PR건에서 제외하고 따로 진행하겠습니다.

ebPark9511 committed:
♻️ Refectoring to XibView

ebPark9511 committed:
♻️ (#130) Refectoring to SYAlert

- Add AlertActionAddinible, AlertActionDefinable, ActionStyleInitiateProtocol
- Add AnimationDefinable, AnimationAlertible
- Add TapDismissibleAlertView

ebPark9511 committed:
✨ (#131) Add basic alert

- (Alertible.swift) Modify show(to:) parameter from call

ebPark9511 committed:
✨ (#132) Add EvaluateOppositeSexAlertView

- (RatingView) Add polymorphism
- Add EvaluateOppositeSexAlertViewTests.swift ,EvaluateOppositeSexAlertView.swift, EvaluateOppositeSexAlertView.xib

ebPark9511 committed:
✨ (#133) Add LowEvaluateOppositeSexAlertView

- Add AssemblyAlertView

ebPark9511 commented:
https://nshipster.com/rawrepresentable/ 를 참고했습니다

Jae-eun commented on Soyeon/Common/SoyeonAlert/AlertActionAddinible.swift:
```diff
@@ -0,0 +1,84 @@
+    func setLayout(to action: DesigndActionType)
+    init(_ style: DesigndActionType, willCompletion closure: @escaping (String?) -> Void)
+    
+    @available(iOS 14.0, *)
```
이거는 왜 14 이상부터 가능한건가요?

ebPark9511 replied:
해당 부분의`init`은 지정할 UI style과 터치 이벤트가 발생했을때 의도한 동작으로 초기화 해주는 역할을 가진 생성자 입니다.
`addTarget`이나 `closure`로 수행될 동작을 추가해야 하게끔 인터페이스를 구성했습니다.

물어보신 부분은 `UIAction`으로 `addTarget`을 호출하는 부분이 버전 제한이 걸려있어서 그렇습니다.

* 기존에는 button에 액션을 추가할때 selector로 추가했어요.
`open func addTarget(_ target: Any?, action: Selector, for controlEvents: UIControl.Event)`

* 그리고 14.0 이후에서는 UIAction으로 액션 추가가 가능해졌습니다.
`func addAction(_ action: UIAction, for controlEvents: UIControl.Event)`
https://developer.apple.com/documentation/uikit/uicontrol/3600490-addaction?language=objc

Jae-eun replied:
답변 감사합니다:)

Jae-eun reviewed (approved):
짝짝 고생하셨습니다

junyng commented on Soyeon/Common/SoyeonAlert/AlertView/AssemblyAlertView.swift:
```diff
@@ -0,0 +1,117 @@
+        return view
+    }()
+    
+    lazy var size: CGSize = .zero {
+        didSet {
+            let fixWidth = contentView.widthAnchor.constraint(equalToConstant: size.width)
+            let height = contentView.heightAnchor.constraint(equalToConstant: size.height)
+            
+            height.priority = .defaultLow
+            
+            NSLayoutConstraint.activate([fixWidth, height])
+        }
+    }
```
`constraints`가 중복으로 적용될 가능성이 있어보여요 
함수로 분리해서 필요할 때 오토레이아웃을 설정하는것이 어떨까요 ?

ebPark9511 replied:
그 부분을 놓쳤었네요 ㅠㅠ 의견 감사합니다!
수정하도록 하겠습니다!

junyng commented on Soyeon/Signup/Step2/WriteProfile1/View/WriteProfileAlert/Action/WriteMyProfileAlertView.swift:
```diff
@@ -0,0 +1,57 @@
+    @IBAction private func didTapCompleteButton(_ sender: Any) {
+        customButtonDelegate?.didTapCompleteButton()
+        
+        if result.isEmpty { return }
```
result string이 빈 값인지에 따라 클로저를 실행할지를 결정하는 것으로 이해했어요! 맞나용 ㅎㅎ?

저는 view에서 비즈니스 로직을 갖고 있어서 어색하게 느껴졌어요 
외부로 부터 주입받아 실행되는 completion 핸들러에서 로직을 수행하는 것이 어떨까요?
```
actionClosure { result in
    if result.isEmpty~
}
```

ebPark9511 replied:
아 맞네요 ㅎㅎ 좋은의견 감사합니다.
수정하도록 하겠습니다.

ebPark9511 replied:
@junyng 님. 해당건은
프로필작성하기1 구조상 외부에서 주입받아 처리하도록 구성하는 부분이 상당히 오래 걸릴듯 합니다..
제가 작업 했었던 부분인데 ^^;; 구조가 엉망이네요..

오래걸리는 이유는 현재 제가 만들어놓은 Alert 구성 protocol과
프로필작성하기1에서 사용되는 알럿의 방식이 매우 상이하기 때문입니다.

이 부분은 개선건으로 이슈보드에 등록해놓고.. 추후에 작업하도록 하겠습니다.

* Data Type
 a. 지정되어 있는 `UIView`의 `viewtag`로 해당하는 입력 스타일의 `enum`을 만듦.
 b. 입력완료는 `(String?) -> Void` 타입으로 고정되어있음.
=> 데이터 특성상 `Int`로 반환 되는 것, `String`으로 반환되는것, `enum`으로 반환되어서 새로 구성되어야 함.
=> `enum` 으로 alert 객체가 `init` 될 수 있음 좋을듯 하다..(이건희망사항)

*  Layout 
연속된 UI가  반복되므로 `UITableView`나 `StackView`의 `addArrangeSubView`로 수정 필요.

junyng replied:
@ebPark9511 넵 알겠습니다~ 👍

junyng reviewed (commented):
alert view가 많아서 설계하는데 어려우셨을텐데 🥲
고생하셨습니다 ~!

ebPark9511 committed:
(#120) Refactor setup rescore layout function

- (AssemblyAlertView) 'A duplicate constraint may occur when changing the size of a variable.' correction
